### PR TITLE
Attention Pattern Matcher (closes #4404)

### PR DESCRIPTION
**Pull request by Fridah-nv**
## Description

Use torch._inductor pattern matcher to match attention

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py [hunk 1]
```diff
@@ -1,105 +1,393 @@
...
+    )
+    return hidden_states.reshape(batch, num_key_value_heads * n_rep, slen, head_dim)
 
-    The pattern is:
-    unsqueeze -> expand -> reshape -> [optional] contiguous
 
-    This is replaced with torch.ops.auto_deploy.torch_attention_repeat_kv.
-    """
-    graph = gm.graph
+def _repeat_kv_pattern2(hidden_states, n_rep) -> torch.Tensor:
```
Comment (Fridah-nv) [inline tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py:26]: @lucaslie Could you comment on why `_repeat_kv_pattern2 ` is needed? I copied this from your branch but did not find use case in my testing.

File: tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py [hunk 2]
```diff
@@ -1,105 +1,393 @@
...
+        value,
+        attn_mask=None,
+        dropout_p=dropout,
+        is_causal=False,
+        scale=scaling,
+    )
+
+
+# no causal_mask, with division, 'complex' model
+def _sfdp_pattern_5(query, key, value, scaling, dropout):
```
Comment (Fridah-nv) [inline tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py:166]: This is to match the [ComplexEagerAttentionModel](https://github.com/nv-auto-deploy/TensorRT-LLM/blob/629174cdf5a1d7f3e090d79a7250f122ad4306b9/tests/unittest/_torch/auto_deploy/unit/singlegpu/transformations/library/test_attention_matcher.py#L260) in the unit test. It does `attn_weights = attn_weights.to(torch.float32)` instead of passing `torch.float32` to `F.softmax`.

@lucaslie do you remember any eager attention implementation that uses this pattern? Llama's eager attention uses `_sfdp_pattern_1` and `_sfdp_pattern_2`.

File: tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py [hunk 3]
```diff
@@ -1,105 +1,393 @@
...
+        op_ignore_types={},
+        scalar_workaround=scalar_workaround,
+    )
+
+    num_matches = patterns.apply(graph)
+    gm = canonicalize_graph(gm)
+    ad_logger.info(f"Found and matched {num_matches} Grouped Attention patterns")
 
 
 def match_causal_attn_mask(gm: GraphModule) -> None:
```
Comment (Fridah-nv) [inline tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py:393]: Should I remove `match_causal_attn_mask ` and its unit test?

File: tests/unittest/_torch/auto_deploy/unit/singlegpu/transformations/library/test_attention_matcher.py [hunk 4]
```diff
@@ -164,14 +165,13 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            # [1, 1, seq_len, seq_len] causal mask with -inf in the upper triangle
+            attn_mask = torch.triu(
+                torch.full((seq_len, seq_len), float("-inf"), device=device), diagonal=1
             )
-            mask = mask.unsqueeze(0).unsqueeze(0)  # [1, 1, seq_len, seq_len]
-            attn_mask = torch.zeros_like(attn_weights, device=device)
-            attn_mask = attn_mask.masked_fill(mask, float("-inf"))
+            attn_mask = (
+                attn_mask.unsqueeze(0).unsqueeze(0).to(x.dtype)
+            )  # shape: [1, 1, seq_len, seq_len]
```
Comment (Fridah-nv) [inline tests/unittest/_torch/auto_deploy/unit/singlegpu/transformations/library/test_attention_matcher.py:174]: The previous mask references `attn_weights` in `attn_mask = torch.zeros_like(attn_weights, device=device)`, this will fail the pattern matcher as it checks num_users and requires the subgraph to be functional and self-contained.

File: tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py [hunk 5]
```diff
@@ -1,105 +1,394 @@
...
+
+def _get_sfdp_patterns() -> List[Dict[str, Any]]:
+    bs = 8
+    seq_len = 16
+    n_heads = 8
+    hidden_size = 512
+    head_dim = hidden_size // n_heads
+    return [
+        {
+            "search_fn": _sfdp_pattern_1,
```
Comment (nvchenghaoz) [inline tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py:221]: Can we add a function like 
```
def create_sfdp_config(search_fn, replace_fn, scaling, dropout, has_attn_mask=False):
     return {"search_fn": search_fn, "replace_fn": replace_fn, ...
     }

config_definitions = [
    # (search_fn,           replace_fn,            scaling,    dropout,    has_attn_mask)
    (_sfdp_pattern_1, _sfdp_replacement_1, 0.1234743, 0.85849734, True),
    (_sfdp_pattern_2, _sfdp_replacement_2, 0.234743,  0.5849734,  False),
]

sfdp_configs = [create_sfdp_config(*config) for config in config_definitions]
```

File: tests/unittest/_torch/auto_deploy/unit/singlegpu/transformations/library/test_attention_matcher.py [hunk 6]
```diff
@@ -139,6 +139,7 @@ def __init__(
         self.out_proj = torch.nn.Linear(num_heads * self.head_dim, hidden_size)
 
     def forward(self, x: torch.Tensor) -> torch.Tensor:
+        print("self.training: ", self.training)
```
Comment (nvchenghaoz) [inline tests/unittest/_torch/auto_deploy/unit/singlegpu/transformations/library/test_attention_matcher.py]: local debug?

**Review by nvchenghaoz — APPROVED** [review]

File: tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py [hunk 7]
```diff
@@ -1,105 +1,394 @@
...
-    """
-    Match and replace the repeat_kv pattern in fx graphs.
+def _repeat_kv_pattern1(hidden_states, n_rep) -> torch.Tensor:
+    batch, num_key_value_heads, slen, head_dim = hidden_states.shape
+    if n_rep == 1:
+        return hidden_states
+    hidden_states = hidden_states[:, :, None, :, :].expand(
+        batch, num_key_value_heads, n_rep, slen, head_dim
+    )
+    return hidden_states.reshape(batch, num_key_value_heads * n_rep, slen, head_dim)
```
[the remark this answers is not in the document]
Comment (Fridah-nv) [reply tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py:23]: I agree that logging which pattern is matched is helpful for debugging and sanity check. Currently `patterns.apply` only returns number of matches for the pass. We may not want to overwrite the apply method in `PatternMatcherPass`. 
One other option is to instantiate different `ADPatternMatcherPass` for each pattern, but then we lose some feature like check pattern duplication. I suggest we verify which pattern is matched in unit tests and avoid registering too many patterns onto one pass. What do you think?

File: tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py (hunk 1 above)
Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py:26]: Probably I forgot to remove it in one of my PRs

File: tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py (hunk 2 above)
Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py:167]: Let's keep it! Seems easy enough to keep it right?

File: tests/unittest/_torch/auto_deploy/unit/singlegpu/transformations/library/test_attention_matcher.py (hunk 4 above)
Comment (lucaslie) [reply tests/unittest/_torch/auto_deploy/unit/singlegpu/transformations/library/test_attention_matcher.py:173]: Is this just a bad unit test or some fundamental aspect we should address?

File: tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py (hunk 2 above)
Comment (Fridah-nv) [reply tensorrt_llm/_torch/auto_deploy/transformations/library/attention.py:167]: I see, thank you

File: tests/unittest/_torch/auto_deploy/unit/singlegpu/transformations/library/test_attention_matcher.py (hunk 4 above)
Comment (Fridah-nv) [reply tests/unittest/_torch/auto_deploy/unit/singlegpu/transformations/library/test_attention_matcher.py:173]: It's just a bad unit test. Just to note that it has stricter constrains than our previous pattern matcher pass. Previously in this case, it will be a match and the subgraph won't be erased completely.

**Comment by Fridah-nv** [comment]
> I am assuming the updated pattern matcher can handle patterns with (GQA) and without repeat_kv (MHA) right?

Correct

**Review by lucaslie — APPROVED** [review]
